adr/00-packages.md

```diff
@@ -0,0 +1,45 @@
+# Architecture Decision Record: Package structure
+
+## Context
+
+We need to specify structure for the application. 
+It will involve choosing how we will divide code into packages and modules. 
+In its current state, application resembles an MVC based architecture. 
+We would like to make a clear decision, what application structure should be going forward.
+
+## Decision
+
+We will divide application into layers represented by packages.
+
+We decided on the following layers:
+* app
+* api
+* infrastructure
+
+The app layer will contain domain models and service interfaces. The api layer will contain endpoint descriptions, without logic.
+Infrastructure layer will house concrete implementations of infrastructure-specific elements. 
+These include database queries based on concrete DB and access library (Quill), and, in the future, possible other technical components like external services, cache, etc.
+Each layer can have separate packages inside. Those packages can group models and services that are connected, by a common entity. 
+For example, User model, UserRepository interface and UserService interface would be placed in the app layer, 
```

Comment:
so what will be the full package structure? I think drawing a ASCII tree might help.

Also, what will the top-level grouping into packages be - using technical, or domain concepts?

Reply:
It would be a cool exercise to divide into domain->technical (where technical means api/app/infra). What domains will we end up with? Just user+article? Should "profile" and "tags" be separate? I recommend the team to analyze the dependencies and try.
Also +1 to represent this as an ASCII tree afterwards.

Reply:
Ultimately I like to divide the "app" into "application logic" and "domain logic", so it would be nice to see if that makes sense here as well. However, putting everything into the "app" (or maybe domain as a better name) seems like a good first step.

Reply:
@adamw I'm not sure if I understand correctly. Do you mean top-level grouping in each layer?

Reply:
@kciesielski 
1. In my current understanding, there would be three main entities, profile, user and article. I would factor profile out as it is currently duplicated by ArticleAuthor and a part of UserData(ProfileData and ArticleAuthor are identical, UserData is just ProfileData with email and token) so that both user and article can use the same underlying model. 
2. We would also move "envelopes" like User, Profile, etc. to API layer and rename them as UserResponse, ProfileResponse, etc. and rename UserData, ProfileData, etc. to just User, Profile and so on
3. We can add a separate value class for Tag so that it distinct from just String. 
4. We can try and separate domain and app. Domain could contain only case classes for entities and app would contain service interfaces and server endpoints with logic implemented using those interfaces.

Reply:
@adamw @kciesielski Is something like this, what you meant by ASCII tree?  
  
app
├── article
│   ├── ArticleRepository
│   ├── ArticleService
│   └── ArticleServerEndpoints
├── profile
│   ├── ProfileRepository
│   ├── ProfileService
│   └── ProfileServerEndpoints
├── tag
│   ├── TagRepository
│   ├── TagService
│   └── TagServerEndpoints
└── user
    ├── UserRepository
    ├── UserService
    └── UserServerEndpoints
api
├── ArticleEndpoints
├── ProfileEndpoints
├── TagEndpoints
└── UserEndpoints
domain
├── Article
├── Profile
├── Tag
└── User
infrastructure
├── QuillArticleRepository
├── QuillProfileRepository
├── QuillTagRepository
└── QuillUserRepository

Reply:
@bartma11 I was thinking about the other way around
core // common domain value classes like UserId, Amount, etc.
├── api
├── domain
├── infrastructure
user
├── api
├── domain
├── infrastructure
article
├── api
├── domain
├── infrastructure
profile
├── api
├── domain
├── infrastructure
tag
├── api
├── domain
├── infrastructure

but this only makes sense if top-level "modules" are pretty independent on each other (which I think aren't). Maybe only user and article can be reasonably separated? Also, this may make us end up with bottom-level packages with a single class in each, wouldn't that be a bit of an overkill @adamw ?